vlog4j-parser/src/main/java/org/semanticweb/vlog4j/parser/javacc/JavaCCParserBase.java

```diff
@@ -86,21 +89,50 @@
 
 	public JavaCCParserBase() {
 		this.knowledgeBase = new KnowledgeBase();
+		this.prefixDeclarations = new LocalPrefixDeclarations();
+		this.parserConfiguration = new DefaultParserConfiguration();
 	}
 
-	Constant createIntegerConstant(String lexicalForm) {
-		return Expressions.makeDatatypeConstant(lexicalForm, PrefixDeclarations.XSD_INTEGER);
+	Constant createConstant(String lexicalForm) throws ParseException {
+		try {
+			return Expressions.makeAbstractConstant(prefixDeclarations.absolutize(lexicalForm));
+		} catch (PrefixDeclarationException e) {
+			throw new ParseException(e.getMessage());
+		}
 	}
 
-	Constant createDecimalConstant(String lexicalForm) {
-		return Expressions.makeDatatypeConstant(lexicalForm, PrefixDeclarations.XSD_DECIMAL);
+	Constant createConstant(String lexicalForm, String datatype) throws ParseException {
+		return createConstant(lexicalForm, null, datatype);
 	}
 
-	Constant createDoubleConstant(String lexicalForm) {
-		return Expressions.makeDatatypeConstant(lexicalForm, PrefixDeclarations.XSD_DOUBLE);
+
+	/**
+	 * Creates a suitable {@link Constant} from the parsed data.
+	 *
+	 * @param string      the string data (unescaped)
+	 * @param languageTag the language tag, or null if not present
+	 * @param datatype    the datatype, or null if not provided
+	 * @return suitable constant
+	 */
+	Constant createConstant(String lexicalForm, String languageTag, String datatype) throws ParseException {
+		try {
+			return parserConfiguration.parseConstant(lexicalForm, languageTag, datatype);
+		} catch (ParsingException e) {
+			ParseException parseException = new ParseException(e.getMessage());
+			parseException.initCause(e);
+			throw parseException;
+		}
 	}
 
-	void addDataSource(String predicateName, int arity, DataSource dataSource) {
+	void addDataSource(String predicateName, int arity, DataSource dataSource) throws ParseException {
+		if (dataSource.getRequiredArity().isPresent()) {
+			Integer requiredArity = dataSource.getRequiredArity().get();
+			if (requiredArity != arity) {
```

Comment:
I know requiredArity cannot be null because Optional.of(null) throws NPE, 
but it is more standard to compare
integer==Integer, instead of Integer == integer.